### PR TITLE
162633510 Ft tag articles

### PR DESCRIPTION
**Pull request by mubstimor**
### What does this PR do?
* Enable authors to add tags to their articles

#### Tasks to be completed
* Add tags to articles being created

#### How should this be manually tested
* Set environment variables in  a `.env` file inside the project's root folder
```
SECRET_KEY='xxxx'
DATABASE_URL=postgres://dbuser:password@localhost:5432/database_name
EMAIL_HOST=xxxx
EMAIL_HOST_USER = xxxx
EMAIL_HOST_PASSWORD = xxxx
EMAIL_PORT=587
EMAIL_USE_TLS = True
TWITTER_CONSUMER_API_KEY=xxxx
TWITTER_CONSUMER_API_SECRET_KEY=xxxx
```
* Run migrations
```
python manage.py makemigrations
python manage.py migrate
```
* Start server
```
python manage.py runserver
```
* Running tests
```
python manage.py test
```
* Register author to be used for creating article
```
POST /api/v1/users/register
{
  "user":{
    "username": "user",
    "email": "johndoe@example.com",
    "password": "pass123"
  }
}

Verify user with email then

POST api/v1/users/login/
{
	"user":{
		"email": "johndoe@example.com",
		"password": "pass123"
	}
}
 Copy the received token to use as a Bearer Token for future requests
```
* Create an article
```
POST /api/v1/articles

Post article with Json request
{
    "title":"Test article",
    "description":"Testing article created yesterday",
    "body":"This is a lorem ipsum section.",
    "tagList": ["Sample", "Tag2", "Nature", "Travel", "BeautifulSites"]
}
```
* View articles with
```
GET /api/v1/articles
```

#### Relevant PT stories
[#162633510](https://www.pivotaltracker.com/story/show/162633510)

### CODE REVIEW

File: authors/apps/articles/urls.py [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
-                    RetrieveUpdateDestroyArticle, RetrieveAuthorArticles)
+                    RetrieveUpdateDestroyArticle, RetrieveAuthorArticles,
+                    ListTag)
 
 app_name = "articles"
 urlpatterns = [
     path('', ListCreateArticle.as_view(), name="articles_list"),
     path('<slug:slug>/', RetrieveUpdateDestroyArticle.as_view(), name="article_detail"),
-    path('author/<int:pk>/', RetrieveAuthorArticles.as_view(), name="author_articles"),
+    path('author/<str:username>/', RetrieveAuthorArticles.as_view(), name="author_articles"),
```
Comment (jeanjoe) [inline authors/apps/articles/urls.py]: Reduce this line to less than 80 characters

File: authors/apps/articles/admin.py [hunk 2]
```diff
@@ -1,12 +1,17 @@
 from django.contrib import admin
 
-from .models import Article, ArticleImage
+from .models import Article, ArticleImage, Tag
 
 class ArticleImageInline(admin.TabularInline):
     model = ArticleImage
     extra = 3
 
+# class TagInline(admin.TabularInline):
```
Comment (jeanjoe) [inline authors/apps/articles/admin.py]: Remove commented code not useful

File: authors/apps/articles/renderers.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+                'article': data
+            })
+
+            
+        except Exception as identifier:
+            pass
+
+        
+        
+
```
Comment (jeanjoe) [inline authors/apps/articles/renderers.py]: Also remove extra line from your work

File: authors/apps/articles/serializers.py [hunk 4]
```diff
@@ -1,24 +1,64 @@
...
+        return value.tag
 
 
 class ArticleSerializer(serializers.ModelSerializer):
-    author = serializers.PrimaryKeyRelatedField(
-        read_only=True, default=serializers.CurrentUserDefault()
-    )
+
+    author = UserSerializer(required=False)
+    # author = ProfileSerializer(read_only=True)
```
Comment (jeanjoe) [inline authors/apps/articles/serializers.py]: remove this commented line of code

**Review by jeanjoe — DISMISSED** [review]

**Comment by jeanjoe** [comment]
LGTM

File: authors/apps/articles/tests/test_articles.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 from django.test import TestCase
-from  rest_framework.reverse import reverse
+from rest_framework.reverse import reverse
 from rest_framework import status
+from pprint import pprint
```
Comment (jeanjoe) [inline authors/apps/articles/tests/test_articles.py]: Remove this unused import

**Review by jeanjoe — CHANGES_REQUESTED** [review]

**Review by jeanjoe — APPROVED** [review]

**Comment by jeanjoe** [comment]
LGTM

**Comment by NabunyaLilian** [comment]
LGTM

**Comment by NabunyaLilian** [comment]
Though you should reduce the number of your commits to one

File: authors/apps/articles/admin.py [hunk 6]
```diff
@@ -1,12 +1,14 @@
...
     model = ArticleImage
     extra = 3
 
+
 class ArticleAdmin(admin.ModelAdmin):
-    inlines = [ ArticleImageInline, ]
+    inlines = [ ArticleImageInline ]
 
 admin.site.register(Article, ArticleAdmin)
+admin.site.register(Tag)
```
Comment (NabunyaLilian) [inline authors/apps/articles/admin.py]: please add a new line

File: authors/apps/articles/serializers.py [hunk 7]
```diff
@@ -1,21 +1,52 @@
...
+
+        return tag[0].id
+
+    def to_representation(self, value):
+        """
+        Serialize tagged objects to a simple textual representation.
+        """
+        return value.tag
+
 from .models import Article, FavoriteArticle
```
Comment (jeanjoe) [inline authors/apps/articles/serializers.py]: Please keep all the imports at the top

**Review by jeanjoe — CHANGES_REQUESTED** [review]

File: authors/apps/articles/serializers.py [hunk 8]
```diff
@@ -25,6 +56,15 @@ class Meta:
         model = Article
         read_only_fields = ('author',)
 
+    def create(self, validated_data):
+        tag = validated_data['tagList']
```
Comment (jeanjoe) [inline authors/apps/articles/serializers.py]: Please remove unused variable

File: authors/apps/articles/serializers.py [hunk 9]
```diff
@@ -1,21 +1,52 @@
...
+
+class TagRelatedField(serializers.RelatedField):
+
+    def get_queryset(self):
+        return Tag.objects.all()
+
+    def to_internal_value(self, data):
+        tag = Tag.objects.get_or_create(tag=data, slug=slugify(data))
+
+        return tag[0].id
```
Comment (jeanjoe) [inline authors/apps/articles/serializers.py]: Please return an object to avoid data obstruction.

**Review by jeanjoe — DISMISSED** [review]

**Comment by Qadriyah** [comment]
LGTM

**Comment by mycok** [comment]
LGTM

**Comment by fatukunda** [comment]
LGTM

**Comment by mubstimor** [comment]
> Though you should reduce the number of your commits to one

noted, thank you!

**Pull request merged**